### ISSUE TITLE
Feature proposal: Partial static functions for instance search and metaprogramming

### ISSUE THREAD

**Issue by jespercockx** [agda/agda#6378]
While working on https://github.com/agda/agda/pull/6368, it got me thinking on the computational model of instance search. Why do we rely on Prolog-style search for instance search, rather than following a more functional-style computation? Instead, we could describe instance search as a **rewrite system** that computes the appropriate instance from the type. For example, if we have a class `Functor` then we could define a "function" `getFunctorInstance : (F : A -> A) -> Functor A` that takes a type as instance and returns the appropriate instance. The behaviour of this "function" could be specified via rewrite rules, e.g. `getFunctorInstance Maybe --> record { fmap = mapMaybe }`. To do recursive instance search, we can then just make a recursive call, e.g. `getMonoidInstance (A x B) --> mkProductMonoid (getMonoidInstance A) (getMonoidInstance B)`.

I am writing "function" between quotes here because there are several reasons why it is not a proper function:
- It is not total, since not every `F : A -> A` has a `Functor` instance.
- In case the instance search is recursive, it might not terminate (the current instance search in Agda uses fuel with a default value of 50).
- It matches on types, which is normally not allowed since types are erased during compilation.

What makes the current incarnation of instance search work despite these issues is the fact that it is run eagerly during type checking, so all instances are resolved *before* the step in compilation where types are erased. But there is no reason that this should be tied to Prolog-style search! How do we define a functional version of this mechanism?

To do this, we need a concept of a **static function**, i.e. a function that is inlined eagerly during type checking (I got this idea from a discussion with @gallais at the Agda meeting). These static functions could do things that are normally not allowed: they could be partial/non-terminating, and they could match on types and other erased arguments. Then, when a static function fails to evaluate statically, this is considered a type error, so static functions do not appear in the run-time code. We could give some leeway here and postpone evaluation of a static function in case it is blocked on a meta-variable, but this postponed evaluation should be solved before we do actual code generation.

The other requirement on these functions is that they should be **extensible** with additional computation rules, in order to enable user code to define its own instances. This makes them very similar to *type families* in Haskell, except that they work for any terms, not just types. In a way, you can already do this in current Agda by postulating a function and extending it with additional rewrite rules on-the-fly. However, this relies on two unsafe features (postulates and rewrite rules), so it would be nicer to package this up in a safe and user-friendly piece of syntax.

One nice side effect of this would be that it guarantees uniqueness of instances by construction since functions are, well, functional. In case there are overlapping instances, the rewrite rules would also overlap, but luckily we already have a confluence checker that ensures the result does not depend on which rewrite rules is chosen.

So far this is just a wild idea, but it seems like this could perhaps be a nice tool to do many things for which we currently use either instance search or reflection; while having a much clearer operational semantics than either of those features. I'm happy to hear any suggestions or criticisms for why this wouldn't work though!

**Comment by anuyts** [comment]
I've been having vague ideas about how typeclasses/instances in Menkar should work and I think you've just nailed what I wanted. A few remarks:
* Currently, instance arguments behave specially when calling a function (they trigger instance resolution) but also when implementing the function (the argument is available as an instance). While I've never been sure whether it makes sense to couple these two behaviours, I do think it's important to be able to declare, somehow, that a function argument should become the output of a given static-function-call.
* One thing that static functions might be able to do, that you haven't mentioned, is to make entirely syntactic distinctions (on normal forms, I guess). An application in HoTT would be
  ```
  getPath : (A : Set) (x y : A) -> Path A x y
  getPath _ (f x) (g y) ---> pathApplicative (getPath _ f g) (getPath _ x y)
  ```
* I don't think static functions should have unique outputs. Generally, you will want a type error when the output is not unique (after deeply expanding all static function calls) (in which case it should be possible to specify the desired value yourself), but sometimes you might want to allow overlapping. For example, the following rule in HoTT could get priority:
  ```
  getPath A x y ---> (getProofThatTypeIsHProp A) x y
  ```
  but of course it should not get in your way if `A` happens to be not an hProp.
* By analogy with the `instance` block (which I think shouldn't be a block but an annotation), I would suggest to state static rewrite rules as annotations, e.g.
  ```
  @(resolves getMonoidInstance (A x B))
  mkProductMonoid : 
    @(resolves getMonoidInstance A)
    Monoid A -> 
    @(resolves getMonoidInstance B)
    Monoid B ->
    Monoid (A x B)
  mkProductMonoid = {!...!}
  ```

I guess I've just re-proposed instance search, but now with some guidance that will keep Agda from considering a monoid instance when a functor instance is required, so the search space is much smaller.

**Comment by jespercockx** [comment]
> Currently, instance arguments behave specially when calling a function (they trigger instance resolution) but also when implementing the function (the argument is available as an instance). While I've never been sure whether it makes sense to couple these two behaviours, I do think it's important to be able to declare, somehow, that a function argument should become the output of a given static-function-call.

Yes, that's an important thing that I haven't yet considered for this proposal. One way to make it work is by having some kind of local rewrite rules.

> One thing that static functions might be able to do, that you haven't mentioned, is to make entirely syntactic distinctions (on normal forms, I guess). An application in HoTT would be [...]

Yes, this would be another thing to explore. I guess the main question is how much reduction you'd want to do during the matching process, with less reduction corresponding to more syntactic matching but also more risk of unpredictability.

> I don't think static functions should have unique outputs. Generally, you will want a type error when the output is not unique (after deeply expanding all static function calls) (in which case it should be possible to specify the desired value yourself), but sometimes you might want to allow overlapping.

Right, that's a valid consideration. Note that the confluence checker is currently optional, so there'd be no immediate problem. If we ever make the confluence checker enabled by default, we should provide a `{-# CONFLUENT #-}` (or {-# NON_CONFLUENT #-}`) pragma to exclude a symbol from the confluence checker (actually, that might be useful to add already now).

**Comment by uhbif19** [comment]
I did not understand idea about instance search, but as for static partial functions, could not it be covered by single built-in function like `staticUnwrapMaybe`?
